### PR TITLE
Factorize code of `ServerEvaluationStrategy` classes, to use the Custom strategy as the basis of other strategies

### PR DESCRIPTION
**Pull request by davidfestal**
### What does this PR do?

This PR makes the `CustomServerEvaluationStrategy` the basis of all other strategies. This allows having the `LocalDockerSinglePortStrategy` simply use a custom template. 

For the moment the properties in `che.properties` are not changed, and other strategies extend the `custom` strategy with specific settings. So this change should have *no impact* on existing configurations / installations.

### What issues does this PR fix or reference?

https://issues.jboss.org/browse/CHE-277

### CODE REVIEW

**Comment by davidfestal** [comment]
@benoitf @l0rd could you review this PR please ?

**Comment by benoitf** [comment]
@davidfestal do you have checked https://github.com/codenvy/codenvy/blob/master/plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/CodenvyDockerServerEvaluationStrategy.java is still compiling ?

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java [hunk 1]
```diff
@@ -72,22 +87,70 @@
...
+     */
+    private boolean throwOnUnknownHost = true;
+    
     /**
      * Default constructor
      */
     @Inject
-    public CustomServerEvaluationStrategy(@Nullable @Named("che.docker.ip") String cheDockerIp,
-                                          @Nullable @Named("che.docker.ip.external") String cheDockerIpExternal,
+    public CustomServerEvaluationStrategy(@Nullable @Named("che.docker.ip") String internalAddress,
```
Comment (benoitf) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java]: maybe it's better to keep the name cheDockerIp / cheDockerIpExternal ?

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java [hunk 2]
```diff
@@ -96,13 +159,17 @@ public CustomServerEvaluationStrategy(@Nullable @Named("che.docker.ip") String c
...
         // get current ports
         Map<String, List<PortBinding>> ports = containerInfo.getNetworkSettings().getPorts();
-
-        return ports.keySet().stream()
-                    .collect(Collectors.toMap(portKey -> portKey,
-                                              portKey -> renderingEvaluation.render(cheDockerCustomExternalTemplate, portKey)));
+        
+        if (isNullOrEmpty(cheDockerCustomExternalTemplate)) {
+            return getExposedPortsToAddressPorts(renderingEvaluation.getExternalAddress(), ports, false);
+        } else {
```
Comment (sunix) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java]: Is this else necessary ?

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java [hunk 3]
```diff
@@ -72,22 +87,70 @@
...
+        this.externalAddressProperty = externalAddress;
         this.chePort = chePort;
         this.cheDockerCustomExternalTemplate = cheDockerCustomExternalTemplate;
         this.cheDockerCustomExternalProtocol = cheDockerCustomExternalProtocol;
+        this.useContainerAddress = useContainerAddress;
     }
 
+    @Override
+    protected Map<String, String> getInternalAddressesAndPorts(ContainerInfo containerInfo, String internalHost) {
+        String internalAddressContainer = containerInfo.getNetworkSettings().getIpAddress();
```
Comment (benoitf) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java]: is it final objects ?

**Review by benoitf — DISMISSED** [review]

**Comment by benoitf** [comment]
@garagatyi ?
@davidfestal as long as it works for codenvy strategy, +1 for me

**Comment by davidfestal** [comment]
> @davidfestal do you have checked https://github.com/codenvy/codenvy/blob/master/plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/CodenvyDockerServerEvaluationStrategy.java is still compiling ?

Good point. I didn't check that. Let me have a look.

**Review by sunix — CHANGES_REQUESTED** [review]

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java [hunk 4]
```diff
@@ -202,12 +274,20 @@ protected OnlineRenderingEvaluation withInternalHost(String internalHost) {
...
-                   gatewayAddressContainer :
-                   this.internalHost;
+        public String getExternalAddress() {
+            if (useContainerAddress) {
+                return externalAddressProperty != null ?
+                    externalAddressProperty :
+                    !isNullOrEmpty(gatewayAddressContainer) ?
+                    gatewayAddressContainer :
+                    this.internalHost;
+            } else {
```
Comment (sunix) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java]: is this `else` necessary ?

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java (hunk 2 above)
Comment (davidfestal) [reply plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java]: > Is this else necessary ?

Sure it isn't. Just for style ;-)
Let me change it.

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java [hunk 5]
```diff
@@ -96,13 +159,17 @@ public CustomServerEvaluationStrategy(@Nullable @Named("che.docker.ip") String c
 
         // create Rendering evaluation
         RenderingEvaluation renderingEvaluation = getOnlineRenderingEvaluation(containerInfo, internalHost);
-
+
```
[the remark this answers is not in the document]
Comment (davidfestal) [reply plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java]: OK. Not hard to remove !

**Comment by davidfestal** [comment]
@sunix is it OK for you now ?

**Comment by davidfestal** [comment]
@benoitf after commit https://github.com/eclipse/che/pull/5366/commits/70cb99b00f1ffb966f45cf2e2961a72403ae9dfa the file [CodenvyDockerServerEvaluationStrategy.java](https://github.com/codenvy/codenvy/blob/master/plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/CodenvyDockerServerEvaluationStrategy.java) should compile fine now (very simple to see it from the code since it directly extends `ServerEvaluationStrategy`). However I can checkout the codenvy repo and test it if you want.

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java [hunk 6]
```diff
@@ -82,12 +108,49 @@ public CustomServerEvaluationStrategy(@Nullable @Named("che.docker.ip") String c
...
                                           @Nullable @Named("che.docker.server_evaluation_strategy.custom.external.protocol") String cheDockerCustomExternalProtocol,
                                           @Named("che.port") String chePort) {
-        super(cheDockerIp, cheDockerIpExternal);
+        this(cheDockerIp, cheDockerIpExternal, cheDockerCustomExternalTemplate, cheDockerCustomExternalProtocol, chePort, false);
+    }
+
+    /**
+     * Constructor to be called by derived strategies
+     */
+    public CustomServerEvaluationStrategy(@Nullable @Named("che.docker.ip") String internalAddress,
```
Comment (garagatyi) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java]: Can you align names of parameters of constructors? E.g. `cheDockerIp == internalAddress`

**Review by garagatyi — APPROVED** [review]

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java (hunk 1 above)
Comment (benoitf) [reply plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java]: @davidfestal I think this comment was not addressed.

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java [hunk 7]
```diff
@@ -302,7 +382,7 @@ protected void initPortMapping() {
         /**
          * Gets default external address.
          */
-        protected String getExternalAddress() {
+        public String getExternalAddress() {
```
Comment (benoitf) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java]: do we need public there ?

Comment (davidfestal) [reply plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java]: yes, because I pulled up the method in the `RenderingEvaluation` interface, since it is now also useful in `CustomServerEvaluationStrategy.getExternalAddressesAndPorts()` (in case `cheDockerCustomExternalTemplate` is `null`)

**Comment by davidfestal** [comment]
@benoitf I checked the compilation of codenvy against this PR. it fails, but not due to this PR. in fact the codenvy build fails with the `openshift-connector-rebased` branch due to this change: https://github.com/eclipse/che/commit/ab740c69b4fecfae9a96c6a64dfb563eee2d7a99#diff-eae0d593d9d11b19a0c27f9e468d957bR73 :
The addition of the `useHttpsForExternalUrls() ` abstract method.

I'll fix this anyway, so we won't have any surprise when merging `openshift-connector-rebased ` to master.

**Comment by benoitf** [comment]
@davidfestal ok, waiting for the fix

Also what about the comments on using variable cheDockerIp instead of internalAddress and cheDockerIpExternal for the externalAddress fields ?
it seems you replaced it in custom server strategy with "old" names

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java [hunk 8]
```diff
@@ -58,6 +58,21 @@
...
 
     /**
+     * Name of the property to get the property that indicates if the machine is the dev machine
+     */
+    public static final String CHE_IS_DEV_MACHINE_PROPERTY = "CHE_IS_DEV_MACHINE=";
+
+    /**
+     * Used to store the address set by property {@code che.docker.ip}, if applicable.
+     */
+    protected String internalAddressProperty;
```
Comment (benoitf) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java]: we should use cheDockerIp

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java [hunk 9]
```diff
@@ -58,6 +58,21 @@
...
+
+    /**
+     * Used to store the address set by property {@code che.docker.ip.external}. if applicable.
+     */
+    protected String externalAddressProperty;
```
Comment (benoitf) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java]: and here cheDockerIpExternal

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java (hunk 8 above)
Comment (davidfestal) [reply plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java]: The names used here are those that were used initially in the `DefaultServerEvaluationStrategy`.

Comment (davidfestal) [reply plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java]: all the other strategies had `internalAddressProperty` and `externalAddressProperty` That's why I had kept them.

Comment (benoitf) [reply plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java]: yes but name of properties have been changed during that time so it's better to keep matching name between the name of the property and the value of the field

**Comment by davidfestal** [comment]
@benoitf I fixed the problem that produce a Codenvy build failure, and changed the names of the custom strategy attributes as requested.

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java [hunk 10]
```diff
@@ -313,14 +393,16 @@ protected String getExternalAddress() {
         protected void populateGlobalProperties() {
             String externalAddress = getExternalAddress();
             String externalIP = getExternalIp(externalAddress);
-            globalPropertiesMap.put("internalIp", internalAddressProperty);
+            globalPropertiesMap.put("internalIp", cheDockerIp);
             globalPropertiesMap.put("externalAddress", externalAddress);
             globalPropertiesMap.put("externalIP", externalIP);
             globalPropertiesMap.put("workspaceId", getWorkspaceId());
+            globalPropertiesMap.put("workspaceIdWithoutPrefix", getWorkspaceId().replaceFirst("workspace",""));
```
Comment (benoitf) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java]: could you move "workspace" as a class constant ?

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java [hunk 11]
```diff
@@ -99,10 +162,14 @@ public CustomServerEvaluationStrategy(@Nullable @Named("che.docker.ip") String c
 
         // get current ports
         Map<String, List<PortBinding>> ports = containerInfo.getNetworkSettings().getPorts();
-
+
```
Comment (sunix) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java]: could this change be removed (trailing space)

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java [hunk 12]
```diff
@@ -99,10 +162,14 @@ public CustomServerEvaluationStrategy(@Nullable @Named("che.docker.ip") String c
 
         // get current ports
         Map<String, List<PortBinding>> ports = containerInfo.getNetworkSettings().getPorts();
-
+        
+        if (isNullOrEmpty(cheDockerCustomExternalTemplate)) {
+            return getExposedPortsToAddressPorts(renderingEvaluation.getExternalAddress(), ports, false);
+        }
+
```
Comment (sunix) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java]: could trailing space be removed ?

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java [hunk 13]
```diff
@@ -99,10 +162,14 @@ public CustomServerEvaluationStrategy(@Nullable @Named("che.docker.ip") String c
...
+        
+        if (isNullOrEmpty(cheDockerCustomExternalTemplate)) {
+            return getExposedPortsToAddressPorts(renderingEvaluation.getExternalAddress(), ports, false);
+        }
+        
         return ports.keySet().stream()
-                    .collect(Collectors.toMap(portKey -> portKey,
-                                              portKey -> renderingEvaluation.render(cheDockerCustomExternalTemplate, portKey)));
+                .collect(Collectors.toMap(portKey -> portKey,
+                                          portKey -> renderingEvaluation.render(cheDockerCustomExternalTemplate, portKey)));
```
Comment (sunix) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java]: if there is no change and just formatting, it would be better if these changes are not included in the PR. (if there is a conflict with another branch it can be a harder to merge). Formatting should be done separately to simplify merge IMO.

**Review by sunix — CHANGES_REQUESTED** [review]

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java (hunk 10 above)
Comment (davidfestal) [reply plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java]: done

Comment (benoitf) [reply plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java]: this property should be also tested in unit test

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java [hunk 14]
```diff
@@ -313,14 +398,16 @@ protected String getExternalAddress() {
...
+            globalPropertiesMap.put("internalIp", cheDockerIp);
             globalPropertiesMap.put("externalAddress", externalAddress);
             globalPropertiesMap.put("externalIP", externalIP);
             globalPropertiesMap.put("workspaceId", getWorkspaceId());
+            globalPropertiesMap.put("workspaceIdWithoutPrefix", getWorkspaceId().replaceFirst(CHE_WORKSPACE_ID_PREFIX,""));
             globalPropertiesMap.put("machineName", getMachineName());
             globalPropertiesMap.put("wildcardNipDomain", getWildcardNipDomain(externalAddress));
             globalPropertiesMap.put("wildcardXipDomain", getWildcardXipDomain(externalAddress));
             globalPropertiesMap.put("chePort", chePort);
+            globalPropertiesMap.put("isDevMachine", getIsDevMachine());
```
Comment (benoitf) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java]: missing unit test

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java [hunk 15]
```diff
@@ -333,12 +420,26 @@ public String render(String template, String port) {
                 this.initialized = true;
             }
             ST stringTemplate = new ST(template);
-            globalPropertiesMap.forEach((key, value) -> stringTemplate.add(key, value));
+            globalPropertiesMap.forEach((key, value) -> stringTemplate.add(key, 
+                                                                           "isDevMachine".equals(key) ?
```
Comment (benoitf) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java]: "isDevMachine" should be a class constant ?

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java (hunk 11 above)
Comment (davidfestal) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java]: Aren't you requesting the exact opposite change as the one you requested [previously](https://github.com/eclipse/che/pull/5366#discussion_r121704027)

File: plugins/plugin-docker/che-plugin-docker-machine/src/test/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategyTest.java [hunk 16]
```diff
@@ -129,6 +130,34 @@ public void testWorkspaceIdRule() throws Throwable {
...
+     */
+    @Test
+    public void testWorkspaceIdWithoutPrefixRule() throws Throwable {
+        this.customServerEvaluationStrategy =
+                new CustomServerEvaluationStrategy("10.0.0.1", "192.168.1.1", "<workspaceIdWithoutPrefix>", "http", "8080");
+
+        Map<String, String> portMapping = this.customServerEvaluationStrategy.getExternalAddressesAndPorts(containerInfo, "localhost");
+
+        Assert.assertTrue(portMapping.containsKey("4401/tcp"));
+        Assert.assertEquals(portMapping.get("4401/tcp"), WORKSPACE_ID_VALUE.replaceFirst(CHE_WORKSPACE_ID_PREFIX, ""));
```
Comment (benoitf) [inline plugins/plugin-docker/che-plugin-docker-machine/src/test/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategyTest.java]: where do we check that we have value ABCDEFG ? for the workspaceId without prefix ?

File: plugins/plugin-docker/che-plugin-docker-machine/src/test/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategyTest.java [hunk 17]
```diff
@@ -129,6 +130,34 @@ public void testWorkspaceIdRule() throws Throwable {
...
+    @Test
+    public void testIsDevMachine() throws Throwable {
+        this.customServerEvaluationStrategy =
+                new CustomServerEvaluationStrategy("10.0.0.1", "192.168.1.1",
+                                                   "<if(isDevMachine)><workspaceId><else><machineName><endif>", "http", "8080");
+
+        Map<String, String> portMapping = this.customServerEvaluationStrategy.getExternalAddressesAndPorts(containerInfo, "localhost");
+
+        Assert.assertTrue(portMapping.containsKey("4401/tcp"));
+        Assert.assertEquals(portMapping.get("4401/tcp"), WORKSPACE_ID_VALUE);
```
Comment (benoitf) [inline plugins/plugin-docker/che-plugin-docker-machine/src/test/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategyTest.java:164]: I don't see where we check the value of the template value ? if we get workspaceId or machineName ?

**Review by benoitf — APPROVED** [review]

**Comment by benoitf** [comment]
@davidfestal @slemeur maybe there is a missing Changelog section in the description

**Comment by davidfestal** [comment]
> @davidfestal @slemeur maybe there is a missing Changelog section in the description

Well, I don't know since these are only internal changes. No property was changed, and the behavior is the same.

**Comment by davidfestal** [comment]
@sunix is it OK now after https://github.com/eclipse/che/pull/5366/commits/cd27b8b589d29426e59d1a15fdb5223f3d327fad ?

**Comment by benoitf** [comment]
@davidfestal yes but template asks for a one-line changelog for all pull requests
(Release notes and docs are optionals for internal refactoring)

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/LocalDockerServerEvaluationStrategy.java [hunk 18]
```diff
@@ -10,19 +10,14 @@
...
 import org.eclipse.che.plugin.docker.client.json.ContainerInfo;
-import org.eclipse.che.plugin.docker.client.json.PortBinding;
 
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
-
-import static com.google.common.base.Strings.isNullOrEmpty;
+import com.google.inject.Inject;
+import com.google.inject.name.Named;
```
Comment (sunix) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/LocalDockerServerEvaluationStrategy.java]: changing order of import is not easy to spot what has been added or removed or not changed

**Comment by davidfestal** [comment]
@benoitf please note that it's a PR against the `openshift-connector-rebased` branch that in turns will be merged to master.
Maybe there will be a changelog for the global PR ?

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/LocalDockerSinglePortServerEvaluationStrategy.java [hunk 19]
```diff
@@ -11,19 +11,11 @@
...
-import org.eclipse.che.plugin.docker.client.json.ContainerInfo;
-import org.eclipse.che.plugin.docker.client.json.PortBinding;
 
-import java.util.Arrays;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
-import java.util.stream.Stream;
+import com.google.inject.Inject;
+import com.google.inject.name.Named;
```
Comment (sunix) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/LocalDockerSinglePortServerEvaluationStrategy.java]: change imports order should be done in a separate PR

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/LocalDockerServerEvaluationStrategy.java (hunk 18 above)
Comment (davidfestal) [reply plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/LocalDockerServerEvaluationStrategy.java]: Obviously due to a "clean imports" in Eclipse.
But let me change this if it blocks your approval.

**Comment by benoitf** [comment]
@davidfestal ah ok I thought it was for master :)

**Comment by benoitf** [comment]
@davidfestal BTW why it isn't directly merged to master ?

**Comment by davidfestal** [comment]
> @davidfestal BTW why it isn't directly merged to master ?

Because the first idea was to have the new `LocalDockerSinglePortServerEvaluationStrategy` (existing only in `openshift-connector` and `openshift-connector-rebased` branches) reuse the new `custom` strategy. But finally, this required some more refactoring. However this started  as a pure Openshift connector change.

**Comment by davidfestal** [comment]
@sunix I fixed the order of the imports

**Comment by benoitf** [comment]
>> @davidfestal BTW why it isn't directly merged to master ?

> Because the first idea was to have the new `LocalDockerSinglePortServerEvaluationStrategy` (existing only in `openshift-connector` and `openshift-connector-rebased` branches) reuse the new `custom` strategy. But finally, this required some more refactoring. However this started  as a pure Openshift connector change.


@davidfestal ok this is because I think that merging into master will allow to decrease the work on merge operations into the openshift-connector-rebased branch, as BTW is being rebased on master.

**Comment by davidfestal** [comment]
> @davidfestal ok this is because I think that merging into master will allow to decrease the work on merge operations into the openshift-connector-rebased branch, as BTW is being rebased on master.

Well, the problem is that it depends on previous changes made on the `openshift-connector-branch`, and is a prerequisite for PR https://github.com/eclipse/che/pull/5110 , which also depends on previous `openshift-connector-branch` changes

**Review by slemeur — APPROVED** [review]
Ok for me as it is going to `openshift-connector-rebased`

File: plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/LocalDockerSinglePortServerEvaluationStrategy.java [hunk 20]
```diff
@@ -41,57 +33,13 @@
...
-            addressesAndPorts.put(serverKey, serverURL);
-        }
-
-        return addressesAndPorts;
-    }
-
-    private String getWorkspaceServerName(Map<String, String> labels, String portKey) {
-        ServerConfImpl serverConf = getServerConfImpl(portKey, labels, new HashMap<>());
-        if (serverConf == null) {
-            return "server-" + portKey.split("/",2)[0];
```
Comment (sunix) [inline plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/LocalDockerSinglePortServerEvaluationStrategy.java:84]: OK these lines have been removed, how do we deal with exposed ports without label ? we did a quick fix here https://github.com/eclipse/che/pull/4970 while waiting for the rebase to work on a refactored and cleanner fix. Would preview urls works with non label exposed ports ?

Comment (davidfestal) [reply plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/LocalDockerSinglePortServerEvaluationStrategy.java:84]: Seems to me it is managed here:

https://github.com/davidfestal/che/blob/0f9a64170ee72077099a89f458ad557f1ae38380/plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/CustomServerEvaluationStrategy.java#L384

Comment (sunix) [reply plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/LocalDockerSinglePortServerEvaluationStrategy.java:84]: thanks

**Review by sunix — APPROVED** [review]
